packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -212,6 +212,9 @@ export enum ContainerMessageType {
 	 * See the [IdCompressor README](./id-compressor/README.md) for more details.
```

Comment:
This comment is for whole review (putting it here to allow proper discussion thread).

I believe what you have here is not export / import flow, where arbitrary transformation over file is possible. Instead, you are essentially doing conversion on the fly, in-place, i.e., the fact of loading container with special container runtime results in transformation of content and a new state. And the only transformations allowed are those that transform individual notes irrespective of the rest of container (where node is DDS or data store).

There are couple directions of exploration from here:
- If developer needs arbitrary transformation (let's say merge 2 DDSs into one instance), how would we go about it?
    - Here, we might say - we do not care about incrementality / preservation of IDs. It's a full rewrite where we do not care if most of the content is the same or not.
- If we are fine with only node-to-node transformations, can we do it somehow in-place inside of summarizer client, such that it could produce incremental summary?
- the most complex - both of above, i.e. relatively arbitrary transformations inside of container runtime that result in incremental summary
    - we need to be careful though with "arbitrary" part here. I think we should allow arbitrary transformations, but I believe they could be constrained to - replace a node (with all of its children) to another node. Here node could be DDS, DataStore, or ContainerRuntime (in later case it means - wipe out all data stores and start from scratch creating any arbitrary set of data stores). I think the key here - developer has choices, but small incremental changes are possible.

Reply:
I do not know if it adds clarify, but here is an example of something similar outside of our project:
1. https://www.inkandswitch.com/cambria/
2. https://github.com/stripe/dagon

The problem of changing data schema is very common in industry, and there is a bunch of interesting work out there. I'm not sure work itself is very applicable, but ideas are.

Here is my understanding how it works usually:
1. There is a framework that allows walking / enumerating through DAG of objects (in our case it's CR/DS (Datastores)/DDSs
2. a set of callbacks (or better - classes with methods, such that they can be stateful) that are called for nodes with ability to replace node with some other node.
3. Such callbacks/classes
   - could replace a node with another node. For example, data store with another data store instance.
   - There could be multiple passes over the graph, such that these classes could collect info on first pass, and act on this info on second pass. I.e. imagine that on first pass it collects info about all Directory instances in container, and on second pass it created one SharedTree instance and patches all other DDSs / Datastores that had reference to Directory instance.
   - GC subsystem is invoked at the end to collect nodes that are not referenced any more.

That said, all such systems usually work on DAG. If we count handles, we deal with arbitrary graph, not DAG. And it's not clear if we want to support such arbitrary transformations that could leave dangling handles. We might want to say initially - it should be possible, and we will explore it later, but for now we allow only simplistic transformation - DDS node to DDS node.

I have no way how to do it, this is mostly discussion allowed to see if we can see some interesting next steps on that way - the way to more generalized system in the future but leveraging principles of that system even today to implement in-place incremental substitution of DDS instance.

Let's chat more :)

Reply:
Here is my poor-man's implementation suggestion: imagine that all parent nodes (container runtime, data store) provide a method to enumerate children and replace child with another (new) child, or patch child in-place (like make changes inside of DDS through normal APIs, but this does not result in ops - though we do not need it now, so defer for later implementation of it). These methods are not exposed on any interfaces, only ContainerRuntime (and CRF) have access to them. For now, we will use only leaf node replacement (i.e. replace DataStore child) for our needs, and will expose transformer that walks the tree and asks transformer if a node (DDS) needs to be replaced. Our transformer will look only for specific factory type (Legacy Shared Tree) and will involve abstract method to replace it with another DDS instance.

Something like that.